File: 0001/two_sum.go
package twoSum

// 8ms 54.7% chris
func twoSum(nums []int, target int) []int {
	// 最快时间4ms，区别在于make
	// m := map[int]int{} -- 4ms
	m := make(map[int]int)

	for i, v := range nums {
		check := target - v
		if _, ok := m[check]; ok {
			if m[check] != i {
				return []int{i, m[check]}
			}
		} else {
			m[v] = i
		}
	}

	return []int{}

}

File: 0001/two_sum.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
    # 236ms 35.71% chris
    # rs = []
    # (0..nums.size-1).each do |i|
    #     rs[0] = i
    #     if nums[i+1..-1].index(target-nums[i])   
    #         rs[1] = nums[i+1..-1].index(target-nums[i])+i+1
    #         break
    #     end
    # end  
    # rs
    
    #48ms 89.55% chris
    hash = {}
    nums.each_with_index {|e,i| hash[e]=i}
    nums.each_with_index do |e,i|
        if hash[target-e] && hash[target-e]!=i
           return [i,hash[target-e]] 
        end
    end
    
    # 44ms 100% other
    # hash = {}
    # nums.each_with_index do |num, idx|
    #     check = target - num
    #     if hash[check] && hash[check] != idx
    #         return [idx, hash[check]]
    #     else
    #         hash[num] = idx
    #     end
    # end
end